### PR TITLE
nolintlint: allow to fix //nolint lines

### LINKED ISSUE #1579: nolintlint: invalid linter name after autofix
Related to https://github.com/golangci/golangci-lint/pull/1573

before the fix:
```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret, gocyclo
func wantedErrors(file, short string) (errs []wantedError) {
```

or

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret,gocyclo
func wantedErrors(file, short string) (errs []wantedError) {
```

after the fix:
```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedreto
func wantedErrors(file, short string) (errs []wantedError) {
```

KO

---

before the fix:

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:gocyclo, nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

or
 
```go
// wantedErrors parses expected errors from comments in a file.
//nolint:gocyclo, nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

after the fix:

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedrett
func wantedErrors(file, short string) (errs []wantedError) {
```

KO

### LINKED ISSUE #1580: nolintlint: too many whitespaces removed after autofix
Related to https://github.com/golangci/golangci-lint/pull/1573

The fix removes all whitespaces, the choice to remove all the whitespaces, in this case, seems not right.

---

before the fix:
```go
// wantedErrors parses expected errors from comments in a file.
// nolint:nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

after the fix:
```go
// wantedErrors parses expected errors from comments in a file.
// nolint:nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

OK

---

before the fix:
```go
// wantedErrors parses expected errors from comments in a file.
//    nolint:nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

after the fix:
```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

KO

### LINKED ISSUE #1581: nolintlint: the autofix is skipped when several unsed linters
Related to https://github.com/golangci/golangci-lint/pull/1573

before the fix:

```go
// wantedErrors parses expected errors from comments in a file.
// nolint:gocyclo,nakedret,whitespace
func wantedErrors(file, short string) (errs []wantedError) {
```

or 

```go
// wantedErrors parses expected errors from comments in a file.
// nolint:nakedret,gocyclo,whitespace
func wantedErrors(file, short string) (errs []wantedError) {
```

or

```go
// wantedErrors parses expected errors from comments in a file.
// nolint:gocyclo,whitespace,nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

results:

```
WARN Line 163 has multiple intersecting issues: []result.Issue{result.Issue{FromLinter:"nolintlint", Text:"directive `// nolint:gocyclo,nakedret,whitespace` is unused for linter \"gocyclo\"", Severity:"", SourceLines:[]string{"// nolint:gocyclo,nakedret,whitespace"}, Replacement:(*result.Replacement)(0xc0074e3ec0), Pkg:(*packages.Package)(0xc0012b3b00), LineRange:(*result.Range)(nil), Pos:token.Position{Filename:"test/errchk.go", Offset:4489, Line:163, Column:1}, HunkPos:0, ExpectNoLint:true, ExpectedNoLintLinter:"gocyclo"}, result.Issue{FromLinter:"nolintlint", Text:"directive `// nolint:gocyclo,nakedret,whitespace` is unused for linter \"whitespace\"", Severity:"", SourceLines:[]string{"// nolint:gocyclo,nakedret,whitespace"}, Replacement:(*result.Replacement)(0xc0074e3f20), Pkg:(*packages.Package)(0xc0012b3b00), LineRange:(*result.Range)(nil), Pos:token.Position{Filename:"test/errchk.go", Offset:4489, Line:163, Column:1}, HunkPos:0, ExpectNoLint:true, ExpectedNoLintLinter:"whitespace"}} 
```

no fix applied.

KO

### PR DESCRIPTION
**Pull request by ashanbrown**
This PR adds a fixing capability to nolintlint. It doesn't try to fix everything.

What it does (when --fix is enabled):

* Removes non-specific //nolint directives that are not used.
* Removes a linter from a linter-dispecific //nolint directive if that linter is not used for that line.
* Removes extra leading whitespace from //. nolint directives for any violations on the number of spaces.  We assume the intent was to have a single leading space unless `allow-leading-whitespace` is set to `false`, in which case we assume no leading spaces (i.e. `//nolint`).

What it does not do:

* Offer explanations when explanations for nolint are required but not provided. I suppose it could add a placeholder that the user could replace but that doesn't seem really in the spirit of how the fix option is supposed to work (since the user might check in that placeholder).
* Make non-specific linter directives specific. This could be done but there is a risk that it cannot divine the intent of the user in setting the nolint directive and I'd argue it's best for the user to make this determination on their own.
* Try to remove linters when multiple linters are specified on a line.  This case is not common and creates some trickiness around handling of commas and the case when all linters are removed.  It's simpler just to ignore it.

Note that fixes may leave trailing spaces or empty lines behind. Presumably, this would be cleaned up by gofmt in most cases which could also fix these cases. I'm not sure whether it's really possible or necessary to try to remove such spaces. It would be nice to at least remove empty lines if there is a way to identify if a comment is the only thing on a line, but I'm not sure how to do that just given the AST. One possibility might be to give the fixer a hint to clean up trailing whitespace or remove blank lines. That could potentially be useful for other linters as well.

Fixes #1579
Fixes #1580
Fixes #1581
Related to #1573

### CODE REVIEW

File: test/testdata/fix/in/nolintlint.go [hunk 1]
```diff
@@ -2,12 +2,15 @@
 //config: linters-settings.nolintlint.allow-leading-space=false
 package p
 
+import "fmt"
+
 func nolintlint() {
-	run() // nolint:bob // leading space should be dropped
-	run() //  nolint:bob // leading spaces should be dropped
+	fmt.Println() // nolint:bob // leading space should be dropped
```
Comment (ashanbrown) [inline test/testdata/fix/in/nolintlint.go:8]: just changed this to something that doesn't trigger other linter errors

**Comment by ldez** [comment]
@ashanbrown could you reference the issue(s)? 

https://github.com/golangci/golangci-lint/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+nolintlint

Have you seen https://github.com/golangci/golangci-lint/pull/1573#issuecomment-751480996 ?

**Comment by ashanbrown** [comment]
@ldez This ended up involving some relatively larger changes than anticipated and might take some time to review, so if you'd prefer that we revert #1573 and roll it together with this, I can put up a PR to do that.

My understanding of the issues with #1573 is that it would potentially lead to bad fixes for issues that already existed, which seems unlikely to adversely affect many (if any) users.

File: pkg/result/processors/nolint.go [hunk 2]
```diff
@@ -30,23 +30,27 @@ func (i *ignoredRange) doesMatch(issue *result.Issue) bool {
 		return false
 	}
 
-	// handle possible unused nolint directives
-	// nolintlint generates potential issues for every nolint directive and they are filtered out here
-	if issue.ExpectNoLint {
-		if issue.ExpectedNoLintLinter != "" {
-			return i.matchedIssueFromLinter[issue.ExpectedNoLintLinter]
+	// only allow selective nolinting of nolintlint
```
Comment (ashanbrown) [inline pkg/result/processors/nolint.go:33]: This has been reorganized so that we can do `//nolint:nolintlint` to disable `nolintlint` checking.

**Comment by ldez** [comment]
> One thing this PR does not do is to remove empty linter lists (such as ,,) or trailing , from a list of linters. I'll put that in a subsequent PR.

> This ended up involving some relatively larger changes than anticipated and might take some time to review, so if you'd prefer that we revert #1573 and roll it together with this, I can put up a PR to do that.

I would prefer to revert #1573 and create one PR with all the elements related to the feature.

File: pkg/result/processors/nolint.go [hunk 3]
```diff
@@ -141,19 +146,14 @@ func (p *Nolint) buildIgnoredRangesForFile(f *ast.File, fset *token.FileSet, fil
 
 func (p *Nolint) shouldPassIssue(i *result.Issue) (bool, error) {
 	nolintDebugf("got issue: %v", *i)
-	if i.FromLinter == golinters.NolintlintName {
-		// always pass nolintlint issues except ones trying find unused nolint directives
-		if !i.ExpectNoLint {
```
Comment (ashanbrown) [inline pkg/result/processors/nolint.go:146]: Now we are allowing nolint directives to be excluded by `//nolint:nolintlint`.

**Comment by ashanbrown** [comment]
@ldez I would love it if you could take a look at this re-attempt to add a `//nolint` fixer.  I believe it addresses all the issues you pointed out with the original fixer.  Thank you.

**Review by ldez — CHANGES_REQUESTED** [review]
I confirm that #1579, #1580, #1581 are fixes.

---

### Trailing comma after the fix

<details>
<summary>one unused linter (begin) [OK]</summary>

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:gocyclo,nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

</details>

<details>
<summary>one unused linter (end) [KO]</summary>

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret,gocyclo
func wantedErrors(file, short string) (errs []wantedError) {
```

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret,
func wantedErrors(file, short string) (errs []wantedError) {
```

</details>

<details>
<summary>two unused linter (begin) [OK]</summary>

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:gocyclo,dupl,nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

</details>

<details>
<summary>two unused linter (end) [KO]</summary>

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:gocyclo,dupl,nakedret
func wantedErrors(file, short string) (errs []wantedError) {
```

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:nakedret,
func wantedErrors(file, short string) (errs []wantedError) {
```

</details>

### Trailing two dot after the fix

<details>
<summary>only unused linters [KO]</summary>

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:gocyclo,dupl
func wantedErrors(file, short string) (errs []wantedError) {
```

```go
// wantedErrors parses expected errors from comments in a file.
//nolint:
func wantedErrors(file, short string) (errs []wantedError) {
```

```
test/errchk.go:163:1: directive `//nolint:` should match `//nolint:[:<comma-separated-linters>] [// <explanation>]` (nolintlint)
//nolint:
^
```

</details>

---

The fact to not have a homogeneous behavior depending on the position of the used linter is not expected for me.

In some cases (ex: only unused linters), the auto-fix produces an error

I think that the trailing commas must handle by this PR.

Also when the auto-fix will produce an empty `//nolint:`, I think that the fix must not be applied, and `Replacement{NeedOnlyDelete: true}` can be better here.

**Comment by ashanbrown** [comment]
@ldez Thanks for looking at this more.  I'm concerned trying to completely cleanly handle the cases with multiple linters specified with `//nolint` will make this PR significantly more complicated and I'd prefer to handle that case in a subsequent PR.  I think these issues could be fixed up by hand after running fix.  The `nolintlint` linter itself will catch the failing cases, preventing them from being committed.  I believe that, due to the possibility of conflicting fixes between different linters, we do not guarantee that "fix" always fixes everything.  If leaving the cases you mention as TODO is ok with you, I'd prefer to merge this as is.  Note that I did this work during my xmas break and I likely won't have more time to spend on this for a few weeks.  That said, if you'd care to try to address those issues before merging, please feel welcome.  I'd be happy to review your changes.

**Comment by ldez** [comment]
@ashanbrown could you clean a bit your PR? I think you made a wrong rebase :wink:

**Review by ldez — CHANGES_REQUESTED** [review]
@ashanbrown could you clean a bit your PR? I think you made a wrong rebase :wink:

**Comment by ashanbrown** [comment]
@ldez I think I've fixed it up.  The one change since last time is to just skip trying to fix unused linter issues for directives with multiple linters specified.

**Comment by ashanbrown** [comment]
@ldez Tests are passing now.  I hadn't fixed them correctly after merging the latest code and needed to add another `expected_linter` field.

**Review by ldez — APPROVED** [review]
To clarify the scope of the PR because the scope of the PR has changed since it was open: this PR only fixes the cases where only one linter is defined in the `nolintlint` directive and this only one linter is unused.

**Pull request merged**